src/utils.py

```diff
@@ -93,10 +93,30 @@ def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
     return wrapper
 
 
+def demand_tls_certificates(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: Any, **kwargs: Any) -> Optional[Any]:
+        event, *_ = args
+        logger.debug(f"Handling event: {event}.")
+
+        if charm.config.get("starttls_enabled", True) and not (
+            charm._container.exists(SERVER_KEY) and charm._container.exists(SERVER_CERT)
+        ):
+            logger.debug(f"TLS certificate and private key not ready. defer event {event}.")
+            event.defer()
+
+            charm.unit.status = BlockedStatus("Missing required TLS certificate and private key")
```

Comment:
Can the charm get to this state if the relation is there, but the certificates are not ready?

If yes (afaict it can), we need to set the status to waiting.
- If user interaction is required the status should be blocked (ie when the relation is not there)
- If no user interaction is required the status should be waiting (ie the relation is there but not ready)

Reply:
Yes, it does have a period when integration is built but certificates are not ready yet. I would like to see it as a transient state. The final steady state would be that certificates are ready in the container. The action that can make the charm reach such a state is by integrating with a certificate provider charm. Not sure if I made myself clear, we can discuss it more tomorrow.

Another thing is that I am planning to refactor those validation decorators since more validations are getting piled together and people don't get a first impression about what those validations gonna do when validation fails. I will organize it into another PR.